feat: Add more granular permissions based on tags

novakzaballa opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [ ] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

- Tags can now be added to roles to enhance granularity in permissions.
<img width="656" alt="Screenshot 2024-06-26 at 11 20 13 AM" src="https://github.com/Flagsmith/flagsmith/assets/41410593/6344f154-ba90-4875-ac9b-835c2f5b43af">

## How did you test this code?

- Go to the "users and permissions" tab
- Click the "roles" tab
- Select the project where you have the tags you want to manage.
- Select the tags, and click save/update button

matthewelwell commented on frontend/web/components/modals/CreateRole.tsx:
```diff
@@ -373,6 +371,44 @@ const CreateRole: FC<CreateRoleType> = ({
+        <FormGroup className='mb-5 setting'>
+          <InputGroup
+            title={'Tags'}
+            tooltip={'Select the project where you want to manage your tags'}
```
This looks wrong, we need to provide good context here about what adding tags to a role means. Can we link out to some documentation?

novakzaballa replied:
I added documentation

kyle-ssg commented on frontend/web/components/modals/CreateRole.tsx:
```diff
@@ -373,6 +371,54 @@ const CreateRole: FC<CreateRoleType> = ({
+                    organisationId={organisationId!}
+                    onChange={(p) => {
+                      setProject(p)
+                      getTags(
```
Is there a reason why this is called? seems wrong.

kyle-ssg replied:
Looking at this more closely, it does look like this is unnecessary. AddEditTags uses useGetTagsQuery.

novakzaballa replied:
You're right, removed.

kyle-ssg commented:
I think I understand, so this lets us create a role with permissions but just targets a tagged set of flags?

If so I have further questions that are not indicated by this work:

Presumably only certain permissions are affected by the tags that are selected and the rest aren't?
   - If this is the case we need to clearly show what setting these affects, it should be in the permissions tab too.
   
   
I'm going to assume that this Delete feature,  Create feature, Update feature state, Create change request and Approve change request? 
   - If this is the case, isn't it quite a big assumption that all the permissions are for the same tags? It feels like this should be set for each permission.

novakzaballa commented:
> I think I understand, so this lets us create a role with permissions but just targets a tagged set of flags?
> 
> If so I have further questions that are not indicated by this work, apologies if I'm assuming something wrong:
> 
> **Presumably only certain permissions are affected by the tags that are selected and the rest aren't?**
> 
> * If this is the case we need to clearly show what setting these affects, it should be in the permissions tab clearly showing what selecting tags does.
> 
> **I'm going to assume that this affects Delete feature, Create feature, Update feature state, Create change request and Approve change request?**
> 
> * If this is the case, isn't it quite a big assumption that all the permissions are for the same tags? It feels like this should be set for each permission otherwise you'd have to create multiple roles with the same users/groups to achieve what you want.

This PR aims to implement the backend created for this [issue](https://github.com/Flagsmith/flagsmith/issues/1535), This means that when a feature is tagged, the role to which a user belongs must have the same flag to delete and update a feature.

kyle-ssg commented:
I'm still basing this on assumptions on how this works but, I think either:

-  1: The UI needs to be improved by moving this to the permissions tab and clearly indicating what this does, a user seeing this for the first time would have a lot of questions, I get that there's docs but that doesn't mean we can have our UI be so unobvious.
    - The docs need to contain a lot more information.
   - Only certain permissions are affected by the tags that are selected and the rest aren't?
   - If it works the way I think it does, this means adding a tag to an existing role without knowing what it does is quite dangerous. We need to clearly state what that would do.

- 2: Consider whether these tags should be at the permission level instead, that'd automatically make this more obvious by how the FE would have to work.

novakzaballa commented:
> I'm still basing this on assumptions about how this works, but I think either:
> 
> * The UI needs to be improved by moving this to the permissions tab and clearly indicating what it does. A user seeing this for the first time would have a lot of questions. I understand that there are docs, but that doesn't mean our UI should be so unclear.
>   
>   * The documentation needs to contain much more information.

1. Added more information about the tags in the role.


>   * Are only certain permissions affected by the selected tags while the rest aren't?

2. Only the Delete feature and Update Feature state permissions are affected.

>   * If it works the way I think it does, adding a tag to an existing role without understanding its impact is quite dangerous. We need to clearly state what that would do.

3. Added a better description in the UI and the Docs.

> * Consider whether these tags should be at the permission level instead. This would automatically make it more obvious by how the front end would need to work.

4. Moved the tags into a new Tags tab inside the permissions tab

kyle-ssg commented:
Hmm I don't think we can provide a good solution to this without some rethinking of the API,

We currently have a well defined use case provided to us:
A company has separate sections of their site managed by different teams that are a mix of developers, QAs etc. Whenever features are created they are tagged with the team e.g. "team_a".


Looking through our permissions it seems like the following are relevant for this usecase

- Delete feature: a member of team A should not be able to delete a feature unless it's tagged with "team_a"

- Create Change Request
- Approve Change Request
- Update Feature State

In addition to the above, the team would like to define the tag permissions in isolation. e.g. product owners can approve change requests.

kyle-ssg commented:
As discussed, let's hide irrelevant things when opting into tag based permissions

![image](https://github.com/user-attachments/assets/5977099c-c25f-4d3c-948a-c166b2c7ff2c)

I think we can hide:
- The organisation tab
- Any project in the project tab that isn't the one selected
- Any project and environment features that aren't relevant, including View environment and View Project

novakzaballa commented:
> 
> I think we can hide:
> 
> * The organisation tab

This tab hides when a tag is saved in the role

> * Any project in the project tab that isn't the one selected

Projects that do not have the same tags as the role should not be displayed.

> * Any project and environment features that aren't relevant, including View environment and View Project

  All features that aren't relevant have been hidden.
